core/src/main/kotlin/org/evomaster/core/search/gene/TupleGene.kt

```diff
@@ -0,0 +1,236 @@
+        }
+    }
+
+    companion object {
+        val log: Logger = LoggerFactory.getLogger(TupleGene::class.java)
+    }
+
+    fun getSpecialGene(): Gene? {
+        TODO("Not yet implemented")
+        if (lastElementTreatedSpecially) {
+            return elements.last()
+        }
+        return null
+
+    }
+
+
+    override fun getValueAsPrintableString(
+        previousGenes: List<Gene>,
+        mode: GeneUtils.EscapeMode?,
+        targetFormat: OutputFormat?,
+        extraCheck: Boolean
+    ): String {
+
+        val buffer = StringBuffer()
+
+        //need the name for input and return
+        buffer.append("$name")
+
+        //printout the inputs. See later if a refactoring is needed
+        if (elements.dropLast(1).isNotEmpty()) {
+            buffer.append("(")
+            val s=elements.dropLast(1).map {
+
+                if (it is EnumGene<*> ||
+                    (it is OptionalGene && it.gene is EnumGene<*>) ||
+                    (it is OptionalGene && it.gene is ArrayGene<*> && it.gene.template is EnumGene<*>) ||
+                    (it is OptionalGene && it.gene is ArrayGene<*> && it.gene.template is OptionalGene && it.gene.template.gene is EnumGene<*>) ||
+                    (it is ArrayGene<*> && it.template is EnumGene<*>) ||
+                    (it is ArrayGene<*> && it.template is OptionalGene && it.template.gene is EnumGene<*>)
+                ) {
+                    val i = it.getValueAsRawString()
+                    "${it.name} : $i"
+                } else {
+                    if (it is ObjectGene || (it is OptionalGene && it.gene is ObjectGene)) {
+                        val i = it.getValueAsPrintableString(mode = GeneUtils.EscapeMode.GQL_INPUT_MODE)
+                        " $i"
+                    } else {
+                        if (it is ArrayGene<*> || (it is OptionalGene && it.gene is ArrayGene<*>)) {
+                            val i = it.getValueAsPrintableString(mode = GeneUtils.EscapeMode.GQL_INPUT_ARRAY_MODE)
+                            "${it.name} : $i"
+                        } else {
+                            val mode =
+                                if (ParamUtil.getValueGene(it) is StringGene) GeneUtils.EscapeMode.GQL_STR_VALUE else GeneUtils.EscapeMode.GQL_INPUT_MODE
+                            val i = it.getValueAsPrintableString(mode = mode, targetFormat = targetFormat)
+                            "${it.name} : $i"
+                        }
+                    }
+                }
+
+            }.joinToString(",").replace("\"", "\\\"")
+            //see another way: eg, joinTo(buffer, ", ").toString().replace("\"", "\\\"") to buffer
+            buffer.append(s)
+            buffer.append(")")
+        }
+
+        //printout the return
+        val returnGene = elements.last()
+
+        buffer.append(
+            if (returnGene is OptionalGene) {
+                assert(returnGene.gene is ObjectGene)
```

Comment:
this is assuming we only deal with GQL. but such `TupleGene` can be used also outside of GQL. so need to keep that in mind (ie, this class should not be too tied to GQL)

Reply:
I have generalize it, if not used for graphql it will print it out as an array.